### PR TITLE
Speed improvements and multiple projection networks

### PR DESCRIPTION
**Pull request by mrcoliva**
This pull request contains a range of performance improvements including
- calling the GNN class with batched input
- performance tweaks all over the place (e.g. removing for loops with optimized numpy operations)
- a unit test that allows to monitor execution speeds and where execution time is spent

Additional changes:
- a new unit test covering the observation to features and edge mapping
- adding multiple projection networks to the actor network
- support for the parameter server in the `GraphObserver`

Sorry, this PR is a bit large because all our branches diverged a bit far. All of is has mostly been seen by all of you already.

### CODE REVIEW

File: examples/data_generation.py [hunk 1]
```diff
@@ -10,7 +10,6 @@
 from absl import flags
```
Comment (lesilvan) [inline examples/data_generation.py]: This file has been moved to another place. It makes no sense to move it back.

File: supervised_learning/py_learn_actor_tests.py [hunk 2]
```diff
@@ -5,17 +5,12 @@
 # https://opensource.org/licenses/MIT
```
Comment (lesilvan) [inline supervised_learning/py_learn_actor_tests.py]: Why change all the file?

**Review by lesilvan — CHANGES_REQUESTED** [review]
From my perspective this merge is not yet usefull. First of all, the test examples fail. Second, a lot of old files are merged and new structure/files killed. Third, useless files for tf2gnn are readded, which were removed before.

I would recommend to first merge develop into activate_nitro and then merge back into it (then you have to look out to not loose your changes), but otherwise we kill all tests once more.

**Comment by mrcoliva** [comment]
@lesilvan There have been a few unintended changes made by the automatic merge. I resolved them as far as I know.

File: supervised_learning/BUILD [hunk 3]
```diff
@@ -1,26 +1,10 @@
-py_test(
```
Comment (lesilvan) [inline supervised_learning/BUILD]: This file should not be changed either. It is one reason, why the CI test fails

**Review by lesilvan — CHANGES_REQUESTED** [review]
supervised_learning/BUILD should not be changed by your branch. The rest looks OK for now, but as long as all tests do not run without error, I would not merge it.

File: supervised_learning/BUILD (hunk 3 above)
Comment (mrcoliva) [reply supervised_learning/BUILD]: Fixed.

**Review by lesilvan — APPROVED** [review]

File: bark_ml/library_wrappers/lib_tf2_gnn/gnn_actor_network.py [hunk 4]
```diff
@@ -103,42 +113,32 @@ def __init__(self,
...
 
     self.call_times = []
     self.gnn_call_times = []
 
+  @property  
+  def output_tensor_spec(self):
+    return self._output_tensor_spec
+
   def call(self, observations, step_type=(), network_state=(), training=False):
+    batch_size, feature_len = observations.shape
```
Comment (lesilvan) [inline bark_ml/library_wrappers/lib_tf2_gnn/gnn_actor_network.py:131]: I think this is the problem. As we do not always batch_call this.

**Review by lesilvan — CHANGES_REQUESTED** [review]

**Review by lesilvan — APPROVED** [review]

**Pull request merged**